### PR TITLE
Enable and fix -Wimplicit-fallthrough warnings

### PR DESCRIPTION
**Pull request by ccawley2011**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -162,11 +162,9 @@ if(CMAKE_C_COMPILER_ID MATCHES "^Intel")
     if(CMAKE_HOST_UNIX)
         set(WARNFLAGS -Wall)
         set(WARNFLAGS_MAINTAINER -Wall -Wcheck -Wremarks)
-        set(WARNFLAGS_DISABLE)
```
Comment (mtl1979) [inline CMakeLists.txt]: Don't remove this... It is still useful for disabling other warnings.

**Review by nmoinvaz — COMMENTED** [review]
I don't have a problem with this PR. I have seen it requested in madler/zlib before. Perhaps we will want to keep `WARNFLAGS_DISABLE` though and keep it empty?

File: CMakeLists.txt [hunk 2]
```diff
@@ -270,9 +266,9 @@ endif()
 
 # Apply warning compiler flags
 if(WITH_MAINTAINER_WARNINGS)
-    add_compile_options(${WARNFLAGS} ${WARNFLAGS_MAINTAINER} ${WARNFLAGS_DISABLE})
```
Comment (mtl1979) [inline CMakeLists.txt]: No reason to change this line...

File: CMakeLists.txt [hunk 3]
```diff
@@ -270,9 +266,9 @@ endif()
...
+    add_compile_options(${WARNFLAGS} ${WARNFLAGS_MAINTAINER})
 else()
-    add_compile_options(${WARNFLAGS} ${WARNFLAGS_DISABLE})
```
Comment (mtl1979) [inline CMakeLists.txt]: No reason to change this line...

File: CMakeLists.txt [hunk 4]
```diff
@@ -162,11 +162,9 @@ if(CMAKE_C_COMPILER_ID MATCHES "^Intel")
     if(CMAKE_HOST_UNIX)
         set(WARNFLAGS -Wall)
         set(WARNFLAGS_MAINTAINER -Wall -Wcheck -Wremarks)
-        set(WARNFLAGS_DISABLE)
     else()
         set(WARNFLAGS /Wall)
         set(WARNFLAGS_MAINTAINER /W5)
-        set(WARNFLAGS_DISABLE)
```
Comment (mtl1979) [inline CMakeLists.txt]: No reason to change this line...

File: CMakeLists.txt [hunk 5]
```diff
@@ -184,7 +182,6 @@ elseif(MSVC)
     # /Oi ?
     set(WARNFLAGS /W3)
     set(WARNFLAGS_MAINTAINER /W4)
-    set(WARNFLAGS_DISABLE)
```
Comment (mtl1979) [inline CMakeLists.txt]: No reason to change this line...

File: CMakeLists.txt [hunk 6]
```diff
@@ -199,7 +196,6 @@ elseif(CMAKE_C_COMPILER_ID MATCHES "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
     # Enable warnings in GCC and Clang
     set(WARNFLAGS -Wall)
     set(WARNFLAGS_MAINTAINER -Wextra -Wpedantic)
-    set(WARNFLAGS_DISABLE -Wno-implicit-fallthrough)
```
Comment (mtl1979) [inline CMakeLists.txt:202]: This is pretty much only warranted change in this file.

**Comment by ccawley2011** [comment]
> Perhaps we will want to keep `WARNFLAGS_DISABLE` though and keep it empty?

OK, done.

**Review by nmoinvaz — APPROVED** [review]

**Comment by mtl1979** [comment]
> I don't have a problem with this PR. I have seen it requested in madler/zlib before. Perhaps we will want to keep `WARNFLAGS_DISABLE` though and keep it empty?

Disabling warnings should only be used as temporary measure before fixing the real issue, unless it is compiler bug or limitation. Keeping `WARNFLAGS_DISABLE` makes it easy to use `list(APPEND WARNFLAGS_DISABLE ...)` to add compiler or target dependent flags to disable warnings that are clearly useless, like we see with MinGW gcc.

**Pull request merged**